relentless/optimize/method.py

```diff
@@ -171,30 +183,139 @@ def optimize(self, objective):
+                    x.value = line_res.design_variables[x]
+                next_res = line_res
+
+            #recycle next result
+            cur_res = next_res
+
             iter_num += 1
 
-        return converged
+        return self.has_converged(cur_res)
+
+    @property
+    def max_iter(self):
+        """int: The maximum number of optimization iterations allowed."""
+        return self._max_iter
+
+    @max_iter.setter
+    def max_iter(self, value):
+        if not isinstance(value, int):
+            raise TypeError('The maximum number of iterations must be an integer.')
+        if value < 1:
+            raise ValueError('The maximum number of iterations must be positive.')
+        self._max_iter = value
 
     @property
     def step_size(self):
-        """float: The step size hyperparameter (:math:`\alpha`)."""
+        r"""float: The step size hyperparameter (:math:`\alpha`)."""
         return self._step_size
 
     @step_size.setter
     def step_size(self, value):
         if value <= 0:
-            raise ValueError('The step size must be positive.')
+            raise ValueError('The specified step size must be positive.')
         self._step_size = value
 
+    @property
+    def line_search(self):
+        """:class:`LineSearch`: The line search object used to optimize the step size."""
+        return self._line_search
+
+    @line_search.setter
+    def line_search(self, value):
+        if value is not None and not isinstance(value, LineSearch):
+            raise TypeError('If defined, the line search parameter must be a LineSearch object.')
+        self._line_search = value
+
+#TODO: document
+class LineSearch:
+    """
+    """
+    def __init__(self, abs_tol, max_iter):
+        self.abs_tol = abs_tol
+        self.max_iter = max_iter
+
+    def find(self, objective, start, end):
+        """
+        """
+        ovars = {x: x.value for x in objective.design_variables()}
+
+        # compute normalized search direction
+        d = end.design_variables - start.design_variables
+        max_step = d.norm()
+        d /= max_step
+
+        # compute start and end target values
+        targets = np.array([d.dot(-start.gradient), d.dot(-end.gradient)])
+
+        # check if max step size acceptable
+        if targets[1] > 0 or np.abs(targets[1]) < self.abs_tol:
+            result = end
+
+        # iterate to minimize target
+        else:
+            steps = np.array([0, max_step])
+            iter_num = 0
+            new_target = self.abs_tol
+            print(steps)
+            print(targets)
+            while np.abs(new_target) >= self.abs_tol and iter_num < self.max_iter:
+                # linear interpolation for step size
+                new_step = (steps[0]*targets[1] - steps[1]*targets[0])/(targets[1] - targets[0])
+
+                # adjust variables based on new step size, compute target
+                for x in ovars:
+                    x.value = ovars[x] + new_step*d[x]
```

Comment:
I think this should be:
```suggestion
                    x.value = start.design_variables[x] + new_step*d[x]
```